File: sdks/node/src/domain/core/cyan_script.ts
import type { CyanPluginInput, CyanProcessorInput } from './cyan_script_model.js';
import type { IInquirer } from './inquirer.js';
import type { Cyan } from './cyan.js';
import type { IDeterminism } from './deterministic.js';
import type { PluginOutput } from '../plugin/output.js';
import type { CyanFileHelper } from './fs/cyan_fs_helper.js';
import type { ProcessorOutput } from '../processor/output.js';
import type { ResolverInput } from '../resolver/input.js';
import type { ResolverOutput } from '../resolver/output.js';

interface ICyanTemplate {
  template(inquirer: IInquirer, determinism: IDeterminism): Promise<Cyan>;
}

interface ICyanProcessor {
  process(input: CyanProcessorInput, fileHelper: CyanFileHelper): Promise<ProcessorOutput>;
}

interface ICyanPlugin {
  plugin(input: CyanPluginInput): Promise<PluginOutput>;
}

interface ICyanResolver {
  resolve(input: ResolverInput): Promise<ResolverOutput>;
}

export type { ICyanTemplate, ICyanProcessor, ICyanPlugin, ICyanResolver };

File: sdks/python/domain/core/cyan_script.py
from abc import abstractmethod, ABC

from domain.core.cyan import Cyan
from domain.core.cyan_script_model import CyanExtensionInput, CyanProcessorInput, CyanPluginInput
from domain.core.deterministic import IDeterminism
from domain.core.fs.cyan_fs_helper import CyanFileHelper
from domain.core.inquirer import IInquirer
from domain.plugin.output import PluginOutput
from domain.processor.output import ProcessorOutput


class ICyanTemplate(ABC):
    @abstractmethod
    async def template(self, inquirer: IInquirer, determinism: IDeterminism) -> Cyan:
        pass


class ICyanExtension(ABC):
    @abstractmethod
    async def extension(self, inquirer: IInquirer, determinism: IDeterminism, i: CyanExtensionInput) -> Cyan:
        pass


class ICyanProcessor(ABC):
    @abstractmethod
    async def process(self, i: CyanProcessorInput, file_helper: CyanFileHelper) -> ProcessorOutput:
        pass


class ICyanPlugin(ABC):
    @abstractmethod
    async def plugin(self, i: CyanPluginInput) -> PluginOutput:
        pass
